conftest.py

```diff
@@ -453,6 +281,280 @@ def tracy_profile():
+    matches_user_requested_silicon_arch = partial(eq, tt_arch)
+    available_archs = tuple(filter(matches_user_requested_silicon_arch, available_archs))
+
+    uses_silicon_arch = "silicon_arch_name" in metafunc.fixturenames
+
+    # sanity
+    if is_test_requesting_specific_silicon_archs and not uses_silicon_arch:
+        raise Exception(
+            f"{metafunc.function} requesting a specific silicon target, but doesn't use silicon_arch_name fixture"
+        )
+
+    if uses_silicon_arch:
+        metafunc.parametrize("silicon_arch_name", available_archs)
+        for test_requested_silicon_arch_fixture in test_requested_silicon_arch_fixtures:
+            # The values of these arch-specific fixtures should not be used in
+            # the test function, so use any parameters, like [True]
+            metafunc.parametrize(test_requested_silicon_arch_fixture, [True])
+
+    input_method = metafunc.config.getoption("--input-method")
+    if input_method == "json":
+        json_path = metafunc.config.getoption("--input-path")
+        if not json_path:
+            raise ValueError("Please provide a valid JSON path using --input-path option.")
+        with open(json_path, "r") as f:
+            data = json.load(f)
+        metafunc.parametrize("user_input", [data])
+    elif input_method == "cli":
+        cli_input = metafunc.config.getoption("--cli-input")
+        if not cli_input:
+            raise ValueError("Please provide input using --cli-input option.")
+        metafunc.parametrize("user_input", [[cli_input]])
+
+
+# Report stashing to get outcomes etc
+phase_report_key = pytest.StashKey()
+
+
+@pytest.hookimpl(tryfirst=True, hookwrapper=True)
+def pytest_runtest_makereport(item, call):
+    # execute all other hooks to obtain the report object
+    outcome = yield
+    rep = outcome.get_result()
+
+    # store test results for each phase of a call, which can
+    # be "setup", "call", "teardown"
+    item.stash.setdefault(phase_report_key, {})[rep.when] = rep
+
+
+@pytest.hookimpl(hookwrapper=True)
+def pytest_runtest_teardown(item, nextitem):
+    yield
+    metal_timeout_enabled = item.config.getoption("--metal-timeout")
+    using_xdist = int(os.getenv("PYTEST_XDIST_WORKER_COUNT", "0"))
+
+    if metal_timeout_enabled is not None or using_xdist:
+        report = item.stash[phase_report_key]
+        test_failed = report.get("call", None) and report["call"].failed
+        if test_failed:
+            logger.info(f"In custom teardown, open device ids: {set(item.pci_ids)}")
+            reset_tensix(set(item.pci_ids))
+
+
+# This is overriding the timer setup hook from pytest-timeout
+# If --metal-timeout is passed, we define a new timeout method that spawns a timer process
+# At timeout, the process kills it's parent (the test process) and then itself
+@pytest.hookimpl(tryfirst=True)
+def pytest_timeout_set_timer(item, settings):
+    metal_timeout_enabled = item.config.getoption("--metal-timeout")
```

Comment:
looks good, but damn this looked annoying to figure out lol

Reply:
:)